src/main/flight/kalman.c

```diff
@@ -0,0 +1,168 @@
+    varStruct.zSumVar = varStruct.zSumVar - (varStruct.zWindow[varStruct.windex] * varStruct.zWindow[varStruct.windex]);
+    varStruct.xySumCoVar = varStruct.xySumCoVar - (varStruct.xWindow[varStruct.windex] * varStruct.yWindow[varStruct.windex]);
+    varStruct.xzSumCoVar = varStruct.xzSumCoVar - (varStruct.xWindow[varStruct.windex] * varStruct.zWindow[varStruct.windex]);
+    varStruct.yzSumCoVar = varStruct.yzSumCoVar - (varStruct.yWindow[varStruct.windex] * varStruct.zWindow[varStruct.windex]);
+
+    varStruct.xMean = varStruct.xSumMean * varStruct.inverseN;
+    varStruct.yMean = varStruct.ySumMean * varStruct.inverseN;
+    varStruct.zMean = varStruct.zSumMean * varStruct.inverseN;
+
+    varStruct.xVar = fabsf(varStruct.xSumVar * varStruct.inverseN - (varStruct.xMean * varStruct.xMean));
+    varStruct.yVar = fabsf(varStruct.ySumVar * varStruct.inverseN - (varStruct.yMean * varStruct.yMean));
+    varStruct.zVar = fabsf(varStruct.zSumVar * varStruct.inverseN - (varStruct.zMean * varStruct.zMean));
+    varStruct.xyCoVar = fabsf(varStruct.xySumCoVar * varStruct.inverseN - (varStruct.xMean * varStruct.yMean));
+    varStruct.xzCoVar = fabsf(varStruct.xzSumCoVar * varStruct.inverseN - (varStruct.xMean * varStruct.zMean));
+    varStruct.yzCoVar = fabsf(varStruct.yzSumCoVar * varStruct.inverseN - (varStruct.yMean * varStruct.zMean));
+
+    float squirt;
+    arm_sqrt_f32(varStruct.xVar + varStruct.xyCoVar + varStruct.xzCoVar, &squirt);
+    kalmanFilterStateRate[X].r = squirt * VARIANCE_SCALE;
+
+    arm_sqrt_f32(varStruct.yVar + varStruct.xyCoVar + varStruct.yzCoVar, &squirt);
+    kalmanFilterStateRate[Y].r = squirt * VARIANCE_SCALE;
+
+    arm_sqrt_f32(varStruct.zVar + varStruct.yzCoVar + varStruct.xzCoVar, &squirt);
+    kalmanFilterStateRate[Z].r = squirt * VARIANCE_SCALE;
+}
+
+float kalman_process(kalman_t *kalmanState, float input, float target)
+{
+    //project the state ahead using acceleration
+    kalmanState->x += (kalmanState->x - kalmanState->lastX);
+
+    //figure out how much to boost or reduce our error in the estimate based on setpoint target.
+    //this should be close to 0 as we approach the sepoint and really high the futher away we are from the setpoint.
+    //update last state
+    kalmanState->lastX = kalmanState->x;
+
+    // calculate the error
+    // float errorMultiplier = fabsf(target - kalmanState->x) * kalmanState->s;
+
+    // // give a boost to the setpoint, used to caluclate the filter cutoff, based on the error and setpoint/gyrodata
+
+    // errorMultiplier = constrainf(errorMultiplier * fabsf(1.0f - (target / kalmanState->lastX)) + 1.0f, 1.0f, 50.0f);
+
+    // if (target != 0.0f)
+    // {
+    //     kalmanState->e = fabsf(1.0f - ((target * errorMultiplier) / kalmanState->lastX));
+    // }
+    // else
+    // {
+    //     kalmanState->e = 1.0f;
+    // }
+
+    kalmanState->e = ABS((target - input) * 3) + ABS(input/4);
```

Comment:
hmm your testing this version of the kalman prediction. In the end we decided against this version of it and went with the above commented out code. I've since updated this version a bit more to allow it to handle its little issues. You should rebase off of what i've currently changed. But i'm excited to see this getting at least tested in iNav :)